{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "78704f5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CODE GENERATED BY:\n",
      "     _    _ _                  __  __       _   _                   \n",
      "    / \\  | (_)_ __   __ _     |  \\/  | __ _| |_| |__   _____      __\n",
      "   / _ \\ | | | '_ \\ / _` |    | |\\/| |/ _` | __| '_ \\ / _ \\ \\ /\\ / /\n",
      "  / ___ \\| | | | | | (_| |    | |  | | (_| | |_| | | |  __/\\ V  V / \n",
      " /_/   \\_\\_|_|_| |_|\\__,_|    |_|  |_|\\__,_|\\__|_| |_|\\___| \\_/\\_/  \n",
      "                                                                    \n",
      "\n",
      "CODONS TO AMINO ACIDS: LOADING COMPLETE!\n",
      "PROPERTIES OF AMINO ACIDS: LOADING COMPLETE!\n",
      "\n",
      "  _________\n",
      " /         \\\n",
      " |  /\\ /\\  |\n",
      " |    -    |\n",
      " |  \\___/  |\n",
      " \\_________/\n",
      "\n",
      "PROVIDE DESIRED LENGTH OF DNA: 9\n",
      "\n",
      "ORIGINAL DNA STRING: GGCCTATAT\n",
      "\n",
      "MUTATED DNA STRING: GGCCTATGT\n",
      "\n",
      "DIFFERENCE FOUND AT CODON 3: AUA != ACA\n",
      "\n",
      "NONCONSERVATIVE MUTATION FOUND:\n",
      "AT AMINO ACID 3: Ile -> 3: Thr\n",
      "\n",
      "        #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "       #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-----\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mT\u001b[00m------\u001b[94mT\u001b[00m#\n",
      "    #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "    #\u001b[94mG\u001b[00m-----\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "      #\u001b[91;1mU\u001b[00m-\u001b[91;1mC\u001b[00m#\n",
      "       ##\n",
      "      #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import sys\n",
    "import time\n",
    "\n",
    "class Compare_Codons:\n",
    "    def __init__(self, DNA1, DNA2):\n",
    "        self.DNA1 = DNA1\n",
    "        self.DNA2 = DNA2 \n",
    "    \n",
    "    def RNA_strings(self):\n",
    "        DNA_to_RNA_Changes = str.maketrans({'A' : 'U', 'G' : 'C', 'C' : 'G', 'T': 'A'})\n",
    "        RNA1 = self.DNA1.translate(DNA_to_RNA_Changes)\n",
    "        RNA2 = self.DNA2.translate(DNA_to_RNA_Changes)\n",
    "        return RNA1, RNA2\n",
    "    \n",
    "    def compare_RNA_strings(self, RNA1_string, RNA2_string):\n",
    "        if len(RNA1_string) != len(RNA2_string):\n",
    "            print('The RNA strings are two different lengths!')\n",
    "            return\n",
    "        for i in range(len(RNA1_string) // 3):\n",
    "            codon1 = RNA1_string[i*3:i*3 + 3]\n",
    "            codon2 = RNA2_string[i*3:i*3 + 3]\n",
    "            if codon1 != codon2:\n",
    "                print(f'DIFFERENCE FOUND AT CODON {i + 1}: {codon1} != {codon2}')\n",
    "                print()\n",
    "\t\t\n",
    "        return\n",
    "    \n",
    "    def print_codons(self, RNA_string):\n",
    "        codons = [RNA_string[i:i + 3] for i in range(0, len(RNA_string), 3)]\n",
    "        for i, codon in enumerate(codons, start=1):\n",
    "            print(f'{i}: {codon}')\n",
    "\t    \n",
    "    \n",
    "    def compare_codons(self, RNA1_string, RNA2_string):\n",
    "        if len(RNA1_string) != len(RNA2_string):\n",
    "            print('The RNA strings are two different lengths!')\n",
    "            return\n",
    "\n",
    "        silent_mutations = []\n",
    "        nonsense1_mutations = []\n",
    "        conservative_mutation = []\n",
    "        nonconservative_mutation = []\n",
    "        \n",
    "        for i in range(len(RNA1_string) // 3):\n",
    "            codon1 = RNA1_string[i*3:i*3 + 3]\n",
    "            codon2 = RNA2_string[i*3:i*3 + 3]\n",
    "\n",
    "            amino_acid1 = codons_to_amino_acids.get(codon1, 'Unknown')\n",
    "            amino_acid2 = codons_to_amino_acids.get(codon2, 'Unknown')\n",
    "            property1 = properties_of_amino_acids.get(codon1, 'Unknown')\n",
    "            property2 = properties_of_amino_acids.get(codon2, 'Unknown')\n",
    "\n",
    "            if codon1 != codon2:\n",
    "                if amino_acid1 == amino_acid2:\n",
    "                    silent_mutations.append((i + 1, codon1, codon2))\n",
    "                elif amino_acid2 == 'Stop' and amino_acid1 != 'Stop':\n",
    "                    nonsense1_mutations.append((i + 1, codon1, codon2))\n",
    "                elif property1 == property2:\n",
    "                    conservative_mutation.append((i + 1, amino_acid1, amino_acid2))\n",
    "                elif property1 != property2:\n",
    "                    nonconservative_mutation.append((i + 1, amino_acid1, amino_acid2))\n",
    "\n",
    "        if silent_mutations:\n",
    "            print(\"SILENT MUTATION FOUND:\")\n",
    "            for position, codon1, codon2 in silent_mutations:\n",
    "                print(f\"AT CODON {position}: {codon1} ({amino_acid1}) -> {codon2} ({amino_acid2})\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "        if nonsense1_mutations:\n",
    "            print(\"NONSENSE MUTATIONS FOUND:\")\n",
    "            for position, codon1, codon2 in nonsense1_mutations:\n",
    "                print(f\"AT CODON {position}: {codon1} ({amino_acid1}) -> {position}: {codon2} (Stop))\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "        if conservative_mutation:\n",
    "            print('CONSERVATIVE MUTATION FOUND:')\n",
    "            for position, codon1, codon2 in conservative_mutation:\n",
    "                print(f\"AT AMINO ACID {position}: {codon1} ({property1}) -> {position}: {codon2} ({property2})\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "        if nonconservative_mutation:\n",
    "            print('NONCONSERVATIVE MUTATION FOUND:')\n",
    "            for position, codon1, codon2 in nonconservative_mutation:\n",
    "                print(f\"AT AMINO ACID {position}: {codon1} -> {position}: {codon2}\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "def find_point_mutations(dna1, dna2):\n",
    "    mutations = []\n",
    "    for i in range(len(dna1)):\n",
    "        if dna1[i] != dna2[i]:\n",
    "            mutations.append((i, dna1[i], dna2[i]))\n",
    "    return mutations\n",
    "\n",
    "def visualize_mutations(dna1, dna2, mutations):\n",
    "    rowIndex = 0\n",
    "    Max_iterations = len(dna1)  # Calculate Max_iterations based on the length of dna1\n",
    "    iteration = 0\n",
    "\n",
    "    while iteration < Max_iterations:  # Limit the loop to a maximum number of iterations\n",
    "        rowIndex = rowIndex + 1\n",
    "        if rowIndex == len(ROWS):\n",
    "            rowIndex = 0\n",
    "\n",
    "        if rowIndex == 0 or rowIndex == 9:\n",
    "            print(ROWS[rowIndex])\n",
    "            continue\n",
    "\n",
    "        randomSelection = random.randint(1, 4)\n",
    "        if randomSelection == 1:\n",
    "            leftNucleotide, rightNucleotide = 'A', 'T'\n",
    "        elif randomSelection == 2:\n",
    "            leftNucleotide, rightNucleotide = 'T', 'A'\n",
    "        elif randomSelection == 3:\n",
    "            leftNucleotide, rightNucleotide = 'C', 'G'\n",
    "        elif randomSelection == 4:\n",
    "            leftNucleotide, rightNucleotide = 'G', 'C'\n",
    "\n",
    "        row = ROWS[rowIndex]\n",
    "        for mutation in mutations:\n",
    "            if mutation[0] == rowIndex - 1:\n",
    "                mutated_row = row.replace('{}', mutation[1], 1)\n",
    "                mutated_row = mutated_row.replace('{}', mutation[2], 1)\n",
    "                mutated_row = mutated_row.replace(mutation[1], f\"\\033[91;1m{mutation[1]}\\033[00m\")\n",
    "                mutated_row = mutated_row.replace(mutation[2], f\"\\033[91;1m{mutation[2]}\\033[00m\")\n",
    "                print(mutated_row)\n",
    "            else:\n",
    "                non_mutated_row = row.replace('{}', f\"\\033[94m{leftNucleotide}\\033[00m\")\n",
    "                non_mutated_row = non_mutated_row.replace('{}', f\"\\033[94m{rightNucleotide}\\033[00m\")\n",
    "                print(non_mutated_row)\n",
    "        time.sleep(PAUSE)\n",
    "        iteration += 1  # Increment the iteration counter\n",
    "\n",
    "# Define a function to generate a random DNA string of given length\n",
    "def generate_random_DNA_string(length):\n",
    "    return ''.join(random.choice('ATCG') for _ in range(length))\n",
    "\n",
    "\n",
    "## Script starts from here\n",
    "\n",
    "# define rows\n",
    "ROWS = [\n",
    "    '         ##',\n",
    "    '        #{}-{}#',\n",
    "    '       #{}---{}#',\n",
    "    '      #{}-----{}#',\n",
    "    '     #{}------{}#',\n",
    "    '    #{}------{}#',\n",
    "    '    #{}-----{}#',\n",
    "    '     #{}---{}#',\n",
    "    '      #{}-{}#',\n",
    "    '       ##',\n",
    "    '      #{}-{}#',\n",
    "    '      #{}---{}#',\n",
    "    '     #{}-----{}#',\n",
    "    '     #{}------{}#',\n",
    "    '      #{}------{}#',\n",
    "    '       #{}-----{}#',\n",
    "    '        #{}---{}#',\n",
    "    '         #{}-{}#',\n",
    "]\n",
    "\n",
    "# define two dictionaries\n",
    "codons_to_amino_acids = {\n",
    "    'UUU' : 'Phe',\n",
    "    'UUC' : 'Phe',\n",
    "    'UUA' : 'Leu',\n",
    "    'UUG' : 'Leu',\n",
    "    'CUU' : 'Leu',\n",
    "    'CUC' : 'Leu',\n",
    "    'CUA' : 'Leu',\n",
    "    'CUG' : 'Leu',\n",
    "    'AUU' : 'Ile',\n",
    "    'AUC' : 'Ile',\n",
    "    'AUA' : 'Ile',\n",
    "    'AUG' : 'Met',\n",
    "    'GUU' : 'Val',\n",
    "    'GUC' : 'Val',\n",
    "    'GUA' : 'Val',\n",
    "    'GUG' : 'Val',\n",
    "    'UCU' : 'Ser',\n",
    "    'UCC' : 'Ser',\n",
    "    'UCA' : 'Ser',\n",
    "    'UCG' : 'Ser',\n",
    "    'CCU' : 'Pro',\n",
    "    'CCC' : 'Pro',\n",
    "    'CCA' : 'Pro',\n",
    "    'CCG' : 'Pro',\n",
    "    'ACU' : 'Thr',\n",
    "    'ACC' : 'Thr',\n",
    "    'ACA' : 'Thr',\n",
    "    'ACG' : 'Thr',\n",
    "    'GCU' : 'Ala',\n",
    "    'GCC' : 'Ala',\n",
    "    'GCA' : 'Ala',\n",
    "    'GCG' : 'Ala',\n",
    "    'UAU' : 'Tyr',\n",
    "    'UAC' : 'Tyr',\n",
    "    'UAA' : 'Stop',\n",
    "    'UAG' : 'Stop',\n",
    "    'CAU' : 'His',\n",
    "    'CAC' : 'His',\n",
    "    'CAA' : 'Gln',\n",
    "    'CAG' : 'Gln',\n",
    "    'AAU' : 'Asn',\n",
    "    'AAC' : 'Asn',\n",
    "    'AAA' : 'Lys',\n",
    "    'AAG' : 'Lys',\n",
    "    'GAU' : 'Asp',\n",
    "    'GAC' : 'Asp',\n",
    "    'GAA' : 'Glu',\n",
    "    'GAG' : 'Glu',\n",
    "    'UGU' : 'Cys',\n",
    "    'UGC' : 'Cys',\n",
    "    'UGA' : 'Stop',\n",
    "    'UGG' : 'Trp',\n",
    "    'CGU' : 'Arg',\n",
    "    'CGC' : 'Arg',\n",
    "    'CGA' : 'Arg',\n",
    "    'CGG' : 'Arg',\n",
    "    'AGU' : 'Ser',\n",
    "    'AGC' : 'Ser',\n",
    "    'AGA' : 'Arg',\n",
    "    'AGG' : 'Arg',\n",
    "    'GGU' : 'Gly',\n",
    "    'GGC' : 'Gly',\n",
    "    'GGA' : 'Gly',\n",
    "    'GGG' : 'Gly'\n",
    "}\n",
    "\n",
    "properties_of_amino_acids = {\n",
    "    'GCU' : 'nonpolar',\n",
    "    'GCC' : 'nonpolar',\n",
    "    'GCA' : 'nonpolar',\n",
    "    'GCG' : 'nonpolar',\n",
    "    'GGU' : 'nonpolar',\n",
    "    'GGC' : 'nonpolar',\n",
    "    'GGA' : 'nonpolar',\n",
    "    'GGG' : 'nonpolar',\n",
    "    'UCU' : 'polar',\n",
    "    'UCC' : 'polar',\n",
    "    'UCA' : 'polar',\n",
    "    'UCG' : 'polar',\n",
    "    'ACU' : 'polar',\n",
    "    'ACC' : 'polar',\n",
    "    'ACA' : 'polar',\n",
    "    'ACG' : 'polar',\n",
    "    'UUA' : 'hydrophobic',\n",
    "    'UUG' : 'hydrophobic',\n",
    "    'CUU' : 'hydrophobic',\n",
    "    'CUC' : 'hydrophobic',\n",
    "    'CUA' : 'hydrophobic',\n",
    "    'CUG' : 'hydrophobic',\n",
    "    'AUU' : 'hydrophobic',\n",
    "    'AUC' : 'hydrophobic',\n",
    "    'AUA' : 'hydrophobic',\n",
    "    'GUU' : 'hydrophobic',\n",
    "    'GUC' : 'hydrophobic',\n",
    "    'GUA' : 'hydrophobic',\n",
    "    'GUG' : 'hydrophobic',\n",
    "    'GAU' : 'negative',\n",
    "    'GAC' : 'negative',\n",
    "    'GAA' : 'negative',\n",
    "    'GAG' : 'negative',\n",
    "    'AAA' : 'positive',\n",
    "    'AAG' : 'positive',\n",
    "    'CGU' : 'positive',\n",
    "    'CGC' : 'positive',\n",
    "    'CGA' : 'positive',\n",
    "    'CGG' : 'positive',\n",
    "    'CAU' : 'positive',\n",
    "    'CAC' : 'positive',\n",
    "    'UUU' : 'aromatic',\n",
    "    'UUC' : 'aromatic',\n",
    "    'UAU' : 'aromatic',\n",
    "    'UAC' : 'aromatic',\n",
    "    'UGG' : 'aromatic'\n",
    "\n",
    "}\n",
    "\n",
    "print('CODE GENERATED BY:')\n",
    "\n",
    "print(\"     _    _ _                  __  __       _   _                   \")\n",
    "print(\"    / \\\\  | (_)_ __   __ _     |  \\\\/  | __ _| |_| |__   _____      __\")\n",
    "print(\"   / _ \\\\ | | | '_ \\\\ / _` |    | |\\\\/| |/ _` | __| '_ \\\\ / _ \\\\ \\\\ /\\\\ / /\")\n",
    "print(\"  / ___ \\\\| | | | | | (_| |    | |  | | (_| | |_| | | |  __/\\\\ V  V / \")\n",
    "print(\" /_/   \\\\_\\\\_|_|_| |_|\\\\__,_|    |_|  |_|\\\\__,_|\\\\__|_| |_|\\\\___| \\\\_/\\\\_/  \")\n",
    "print(\"                                                                    \")\n",
    "\n",
    "print()\n",
    "\n",
    "print(\"CODONS TO AMINO ACIDS: LOADING COMPLETE!\")\n",
    "print(\"PROPERTIES OF AMINO ACIDS: LOADING COMPLETE!\")\n",
    "print()\n",
    "print('  _________\\n /         \\\\\\n |  /\\\\ /\\\\  |\\n |    -    |\\n |  \\\\___/  |\\n \\\\_________/')\n",
    "\n",
    "print()\n",
    "\n",
    "DNA_length = int(input(\"PROVIDE DESIRED LENGTH OF DNA: \"))\n",
    "print()\n",
    "\n",
    "PAUSE = 0.15  # Change it to adjust the animation speed\n",
    "\n",
    "\n",
    "\n",
    "try:\n",
    "\n",
    "    # Generate random DNA strings\n",
    "    original_DNA = generate_random_DNA_string(DNA_length)\n",
    "    mutated_DNA = original_DNA[:]   #copies the original DNA so then later we can introduce mutation in a few base pairs\n",
    "\n",
    "    # Introduce mutations in a few base pairs\n",
    "    num_mutations = random.randint(1, 5)\n",
    "    for _ in range(num_mutations):\n",
    "        mutation_position = random.randint(0, len(mutated_DNA) - 1)\n",
    "        mutated_base = random.choice('ATCG'.replace(mutated_DNA[mutation_position], ''))\n",
    "        mutated_DNA = mutated_DNA[:mutation_position] + mutated_base + mutated_DNA[mutation_position + 1:]\n",
    "\n",
    "    print(\"ORIGINAL DNA STRING:\", original_DNA)\n",
    "    print()\n",
    "    print(\"MUTATED DNA STRING:\", mutated_DNA)\n",
    "    print()\n",
    "\n",
    "    compare = Compare_Codons(original_DNA, mutated_DNA)\n",
    "    RNA1, RNA2 = compare.RNA_strings()\n",
    "    compare.compare_RNA_strings(RNA1, RNA2)\n",
    "    compare.compare_codons(RNA1, RNA2)\n",
    "\n",
    "    # Find point mutations and visualize them\n",
    "    mutations = find_point_mutations(RNA1, RNA2)\n",
    "    visualize_mutations(original_DNA, mutated_DNA, mutations)\n",
    "except KeyboardInterrupt:\n",
    "    sys.exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f400951f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CODE GENERATED BY:\n",
      "     _    _ _                  __  __       _   _                   \n",
      "    / \\  | (_)_ __   __ _     |  \\/  | __ _| |_| |__   _____      __\n",
      "   / _ \\ | | | '_ \\ / _` |    | |\\/| |/ _` | __| '_ \\ / _ \\ \\ /\\ / /\n",
      "  / ___ \\| | | | | | (_| |    | |  | | (_| | |_| | | |  __/\\ V  V / \n",
      " /_/   \\_\\_|_|_| |_|\\__,_|    |_|  |_|\\__,_|\\__|_| |_|\\___| \\_/\\_/  \n",
      "                                                                    \n",
      "\n",
      "CODONS TO AMINO ACIDS: LOADING COMPLETE!\n",
      "PROPERTIES OF AMINO ACIDS: LOADING COMPLETE!\n",
      "\n",
      "  _________\n",
      " /         \\\n",
      " |  /\\ /\\  |\n",
      " |    -    |\n",
      " |  \\___/  |\n",
      " \\_________/\n",
      "\n",
      "PROVIDE DESIRED LENGTH OF DNA: 33\n",
      "\n",
      "ORIGINAL DNA STRING: CCGGCTAAGGAAAAAGTATTGGCCCCTCCTCCA\n",
      "\n",
      "MUTATED DNA STRING: CCGGCTATTGAAAAAGTCTTGGCCCCTCCTCCA\n",
      "\n",
      "DIFFERENCE FOUND AT CODON 3: UUC != UAA\n",
      "\n",
      "DIFFERENCE FOUND AT CODON 6: CAU != CAG\n",
      "\n",
      "NONSENSE MUTATIONS FOUND:\n",
      "AT CODON 3: UUC (Gly) -> 3: UAA (Stop))\n",
      "\n",
      "NONCONSERVATIVE MUTATION FOUND:\n",
      "AT AMINO ACID 6: His -> 6: Gln\n",
      "\n",
      "        #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "        #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "        #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "       #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "       #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "       #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "    #\u001b[94mC\u001b[00m------\u001b[94mC\u001b[00m#\n",
      "    #\u001b[94mC\u001b[00m------\u001b[94mC\u001b[00m#\n",
      "    #\u001b[94mC\u001b[00m------\u001b[94mC\u001b[00m#\n",
      "    #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "    #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "    #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "     #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "     #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "     #\u001b[94mC\u001b[00m---\u001b[94mC\u001b[00m#\n",
      "      #\u001b[91;1mU\u001b[00m-\u001b[91;1mA\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "       ##\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m---\u001b[94mA\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m---\u001b[94mA\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m---\u001b[94mA\u001b[00m#\n",
      "     #\u001b[94mA\u001b[00m-----\u001b[94mA\u001b[00m#\n",
      "     #\u001b[94mA\u001b[00m-----\u001b[94mA\u001b[00m#\n",
      "     #\u001b[94mA\u001b[00m-----\u001b[94mA\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m------\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "       #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "       #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "       #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "        #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "        #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "        #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "         #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "         #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "         #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "         ##\n",
      "        #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "        #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "        #\u001b[94mT\u001b[00m-\u001b[94mT\u001b[00m#\n",
      "       #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "       #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "       #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "     #\u001b[94mT\u001b[00m------\u001b[94mT\u001b[00m#\n",
      "     #\u001b[94mT\u001b[00m------\u001b[94mT\u001b[00m#\n",
      "     #\u001b[94mT\u001b[00m------\u001b[94mT\u001b[00m#\n",
      "    #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "    #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "    #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "    #\u001b[94mC\u001b[00m-----\u001b[94mC\u001b[00m#\n",
      "    #\u001b[94mC\u001b[00m-----\u001b[94mC\u001b[00m#\n",
      "    #\u001b[94mC\u001b[00m-----\u001b[94mC\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m---\u001b[94mG\u001b[00m#\n",
      "      #\u001b[91;1mU\u001b[00m-\u001b[91;1mA\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "      #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "       ##\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mG\u001b[00m-\u001b[94mG\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m---\u001b[94mT\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m---\u001b[94mT\u001b[00m#\n",
      "      #\u001b[94mT\u001b[00m---\u001b[94mT\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m-----\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m-----\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mG\u001b[00m-----\u001b[94mG\u001b[00m#\n",
      "     #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "     #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "     #\u001b[94mA\u001b[00m------\u001b[94mA\u001b[00m#\n",
      "      #\u001b[94mC\u001b[00m------\u001b[94mC\u001b[00m#\n",
      "      #\u001b[94mC\u001b[00m------\u001b[94mC\u001b[00m#\n",
      "      #\u001b[94mC\u001b[00m------\u001b[94mC\u001b[00m#\n",
      "       #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "       #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "       #\u001b[94mT\u001b[00m-----\u001b[94mT\u001b[00m#\n",
      "        #\u001b[94mA\u001b[00m---\u001b[94mA\u001b[00m#\n",
      "        #\u001b[94mA\u001b[00m---\u001b[94mA\u001b[00m#\n",
      "        #\u001b[94mA\u001b[00m---\u001b[94mA\u001b[00m#\n",
      "         #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "         #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "         #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "         ##\n",
      "        #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "        #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n",
      "        #\u001b[94mA\u001b[00m-\u001b[94mA\u001b[00m#\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import sys\n",
    "import time\n",
    "\n",
    "class Compare_Codons:\n",
    "    def __init__(self, DNA1, DNA2):\n",
    "        self.DNA1 = DNA1\n",
    "        self.DNA2 = DNA2 \n",
    "    \n",
    "    def RNA_strings(self):\n",
    "        DNA_to_RNA_Changes = str.maketrans({'A' : 'U', 'G' : 'C', 'C' : 'G', 'T': 'A'})\n",
    "        RNA1 = self.DNA1.translate(DNA_to_RNA_Changes)\n",
    "        RNA2 = self.DNA2.translate(DNA_to_RNA_Changes)\n",
    "        return RNA1, RNA2\n",
    "    \n",
    "    def compare_RNA_strings(self, RNA1_string, RNA2_string):\n",
    "        if len(RNA1_string) != len(RNA2_string):\n",
    "            print('The RNA strings are two different lengths!')\n",
    "            return\n",
    "        for i in range(len(RNA1_string) // 3):\n",
    "            codon1 = RNA1_string[i*3:i*3 + 3]\n",
    "            codon2 = RNA2_string[i*3:i*3 + 3]\n",
    "            if codon1 != codon2:\n",
    "                print(f'DIFFERENCE FOUND AT CODON {i + 1}: {codon1} != {codon2}')\n",
    "                print()\n",
    "\t\t\n",
    "        return\n",
    "    \n",
    "    def print_codons(self, RNA_string):\n",
    "        codons = [RNA_string[i:i + 3] for i in range(0, len(RNA_string), 3)]\n",
    "        for i, codon in enumerate(codons, start=1):\n",
    "            print(f'{i}: {codon}')\n",
    "\t    \n",
    "    \n",
    "    def compare_codons(self, RNA1_string, RNA2_string):\n",
    "        if len(RNA1_string) != len(RNA2_string):\n",
    "            print('The RNA strings are two different lengths!')\n",
    "            return\n",
    "\n",
    "        silent_mutations = []\n",
    "        nonsense1_mutations = []\n",
    "        conservative_mutation = []\n",
    "        nonconservative_mutation = []\n",
    "        \n",
    "        for i in range(len(RNA1_string) // 3):\n",
    "            codon1 = RNA1_string[i*3:i*3 + 3]\n",
    "            codon2 = RNA2_string[i*3:i*3 + 3]\n",
    "\n",
    "            amino_acid1 = codons_to_amino_acids.get(codon1, 'Unknown')\n",
    "            amino_acid2 = codons_to_amino_acids.get(codon2, 'Unknown')\n",
    "            property1 = properties_of_amino_acids.get(codon1, 'Unknown')\n",
    "            property2 = properties_of_amino_acids.get(codon2, 'Unknown')\n",
    "\n",
    "            if codon1 != codon2:\n",
    "                if amino_acid1 == amino_acid2:\n",
    "                    silent_mutations.append((i + 1, codon1, codon2))\n",
    "                elif amino_acid2 == 'Stop' and amino_acid1 != 'Stop':\n",
    "                    nonsense1_mutations.append((i + 1, codon1, codon2))\n",
    "                elif property1 == property2:\n",
    "                    conservative_mutation.append((i + 1, amino_acid1, amino_acid2))\n",
    "                elif property1 != property2:\n",
    "                    nonconservative_mutation.append((i + 1, amino_acid1, amino_acid2))\n",
    "\n",
    "        if silent_mutations:\n",
    "            print(\"SILENT MUTATION FOUND:\")\n",
    "            for position, codon1, codon2 in silent_mutations:\n",
    "                print(f\"AT CODON {position}: {codon1} ({amino_acid1}) -> {codon2} ({amino_acid2})\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "        if nonsense1_mutations:\n",
    "            print(\"NONSENSE MUTATIONS FOUND:\")\n",
    "            for position, codon1, codon2 in nonsense1_mutations:\n",
    "                print(f\"AT CODON {position}: {codon1} ({amino_acid1}) -> {position}: {codon2} (Stop))\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "        if conservative_mutation:\n",
    "            print('CONSERVATIVE MUTATION FOUND:')\n",
    "            for position, codon1, codon2 in conservative_mutation:\n",
    "                print(f\"AT AMINO ACID {position}: {codon1} ({property1}) -> {position}: {codon2} ({property2})\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "        if nonconservative_mutation:\n",
    "            print('NONCONSERVATIVE MUTATION FOUND:')\n",
    "            for position, codon1, codon2 in nonconservative_mutation:\n",
    "                print(f\"AT AMINO ACID {position}: {codon1} -> {position}: {codon2}\")\n",
    "                print()\n",
    "\t\t\n",
    "\n",
    "def find_point_mutations(dna1, dna2):\n",
    "    mutations = []\n",
    "    for i in range(len(dna1)):\n",
    "        if dna1[i] != dna2[i]:\n",
    "            mutations.append((i, dna1[i], dna2[i]))\n",
    "    return mutations\n",
    "\n",
    "def visualize_mutations(dna1, dna2, mutations):\n",
    "    rowIndex = 0\n",
    "    Max_iterations = len(dna1)  # Calculate Max_iterations based on the length of dna1\n",
    "    iteration = 0\n",
    "\n",
    "    while iteration < Max_iterations:  # Limit the loop to a maximum number of iterations\n",
    "        rowIndex = rowIndex + 1\n",
    "        if rowIndex == len(ROWS):\n",
    "            rowIndex = 0\n",
    "\n",
    "        if rowIndex == 0 or rowIndex == 9:\n",
    "            print(ROWS[rowIndex])\n",
    "            continue\n",
    "\n",
    "        randomSelection = random.randint(1, 4)\n",
    "        if randomSelection == 1:\n",
    "            leftNucleotide, rightNucleotide = 'A', 'T'\n",
    "        elif randomSelection == 2:\n",
    "            leftNucleotide, rightNucleotide = 'T', 'A'\n",
    "        elif randomSelection == 3:\n",
    "            leftNucleotide, rightNucleotide = 'C', 'G'\n",
    "        elif randomSelection == 4:\n",
    "            leftNucleotide, rightNucleotide = 'G', 'C'\n",
    "\n",
    "        row = ROWS[rowIndex]\n",
    "        for mutation in mutations:\n",
    "            if mutation[0] == rowIndex - 1:\n",
    "                mutated_row = row.replace('{}', mutation[1], 1)\n",
    "                mutated_row = mutated_row.replace('{}', mutation[2], 1)\n",
    "                mutated_row = mutated_row.replace(mutation[1], f\"\\033[91;1m{mutation[1]}\\033[00m\")\n",
    "                mutated_row = mutated_row.replace(mutation[2], f\"\\033[91;1m{mutation[2]}\\033[00m\")\n",
    "                print(mutated_row)\n",
    "            else:\n",
    "                non_mutated_row = row.replace('{}', f\"\\033[94m{leftNucleotide}\\033[00m\")\n",
    "                non_mutated_row = non_mutated_row.replace('{}', f\"\\033[94m{rightNucleotide}\\033[00m\")\n",
    "                print(non_mutated_row)\n",
    "        time.sleep(PAUSE)\n",
    "        iteration += 1  # Increment the iteration counter\n",
    "\n",
    "# Define a function to generate a random DNA string of given length\n",
    "def generate_random_DNA_string(length):\n",
    "    return ''.join(random.choice('ATCG') for _ in range(length))\n",
    "\n",
    "\n",
    "## Script starts from here\n",
    "\n",
    "# define rows\n",
    "ROWS = [\n",
    "    '         ##',\n",
    "    '        #{}-{}#',\n",
    "    '       #{}---{}#',\n",
    "    '      #{}-----{}#',\n",
    "    '     #{}------{}#',\n",
    "    '    #{}------{}#',\n",
    "    '    #{}-----{}#',\n",
    "    '     #{}---{}#',\n",
    "    '      #{}-{}#',\n",
    "    '       ##',\n",
    "    '      #{}-{}#',\n",
    "    '      #{}---{}#',\n",
    "    '     #{}-----{}#',\n",
    "    '     #{}------{}#',\n",
    "    '      #{}------{}#',\n",
    "    '       #{}-----{}#',\n",
    "    '        #{}---{}#',\n",
    "    '         #{}-{}#',\n",
    "]\n",
    "\n",
    "# define two dictionaries\n",
    "codons_to_amino_acids = {\n",
    "    'UUU' : 'Phe',\n",
    "    'UUC' : 'Phe',\n",
    "    'UUA' : 'Leu',\n",
    "    'UUG' : 'Leu',\n",
    "    'CUU' : 'Leu',\n",
    "    'CUC' : 'Leu',\n",
    "    'CUA' : 'Leu',\n",
    "    'CUG' : 'Leu',\n",
    "    'AUU' : 'Ile',\n",
    "    'AUC' : 'Ile',\n",
    "    'AUA' : 'Ile',\n",
    "    'AUG' : 'Met',\n",
    "    'GUU' : 'Val',\n",
    "    'GUC' : 'Val',\n",
    "    'GUA' : 'Val',\n",
    "    'GUG' : 'Val',\n",
    "    'UCU' : 'Ser',\n",
    "    'UCC' : 'Ser',\n",
    "    'UCA' : 'Ser',\n",
    "    'UCG' : 'Ser',\n",
    "    'CCU' : 'Pro',\n",
    "    'CCC' : 'Pro',\n",
    "    'CCA' : 'Pro',\n",
    "    'CCG' : 'Pro',\n",
    "    'ACU' : 'Thr',\n",
    "    'ACC' : 'Thr',\n",
    "    'ACA' : 'Thr',\n",
    "    'ACG' : 'Thr',\n",
    "    'GCU' : 'Ala',\n",
    "    'GCC' : 'Ala',\n",
    "    'GCA' : 'Ala',\n",
    "    'GCG' : 'Ala',\n",
    "    'UAU' : 'Tyr',\n",
    "    'UAC' : 'Tyr',\n",
    "    'UAA' : 'Stop',\n",
    "    'UAG' : 'Stop',\n",
    "    'CAU' : 'His',\n",
    "    'CAC' : 'His',\n",
    "    'CAA' : 'Gln',\n",
    "    'CAG' : 'Gln',\n",
    "    'AAU' : 'Asn',\n",
    "    'AAC' : 'Asn',\n",
    "    'AAA' : 'Lys',\n",
    "    'AAG' : 'Lys',\n",
    "    'GAU' : 'Asp',\n",
    "    'GAC' : 'Asp',\n",
    "    'GAA' : 'Glu',\n",
    "    'GAG' : 'Glu',\n",
    "    'UGU' : 'Cys',\n",
    "    'UGC' : 'Cys',\n",
    "    'UGA' : 'Stop',\n",
    "    'UGG' : 'Trp',\n",
    "    'CGU' : 'Arg',\n",
    "    'CGC' : 'Arg',\n",
    "    'CGA' : 'Arg',\n",
    "    'CGG' : 'Arg',\n",
    "    'AGU' : 'Ser',\n",
    "    'AGC' : 'Ser',\n",
    "    'AGA' : 'Arg',\n",
    "    'AGG' : 'Arg',\n",
    "    'GGU' : 'Gly',\n",
    "    'GGC' : 'Gly',\n",
    "    'GGA' : 'Gly',\n",
    "    'GGG' : 'Gly'\n",
    "}\n",
    "\n",
    "properties_of_amino_acids = {\n",
    "    'GCU' : 'nonpolar',\n",
    "    'GCC' : 'nonpolar',\n",
    "    'GCA' : 'nonpolar',\n",
    "    'GCG' : 'nonpolar',\n",
    "    'GGU' : 'nonpolar',\n",
    "    'GGC' : 'nonpolar',\n",
    "    'GGA' : 'nonpolar',\n",
    "    'GGG' : 'nonpolar',\n",
    "    'UCU' : 'polar',\n",
    "    'UCC' : 'polar',\n",
    "    'UCA' : 'polar',\n",
    "    'UCG' : 'polar',\n",
    "    'ACU' : 'polar',\n",
    "    'ACC' : 'polar',\n",
    "    'ACA' : 'polar',\n",
    "    'ACG' : 'polar',\n",
    "    'UUA' : 'hydrophobic',\n",
    "    'UUG' : 'hydrophobic',\n",
    "    'CUU' : 'hydrophobic',\n",
    "    'CUC' : 'hydrophobic',\n",
    "    'CUA' : 'hydrophobic',\n",
    "    'CUG' : 'hydrophobic',\n",
    "    'AUU' : 'hydrophobic',\n",
    "    'AUC' : 'hydrophobic',\n",
    "    'AUA' : 'hydrophobic',\n",
    "    'GUU' : 'hydrophobic',\n",
    "    'GUC' : 'hydrophobic',\n",
    "    'GUA' : 'hydrophobic',\n",
    "    'GUG' : 'hydrophobic',\n",
    "    'GAU' : 'negative',\n",
    "    'GAC' : 'negative',\n",
    "    'GAA' : 'negative',\n",
    "    'GAG' : 'negative',\n",
    "    'AAA' : 'positive',\n",
    "    'AAG' : 'positive',\n",
    "    'CGU' : 'positive',\n",
    "    'CGC' : 'positive',\n",
    "    'CGA' : 'positive',\n",
    "    'CGG' : 'positive',\n",
    "    'CAU' : 'positive',\n",
    "    'CAC' : 'positive',\n",
    "    'UUU' : 'aromatic',\n",
    "    'UUC' : 'aromatic',\n",
    "    'UAU' : 'aromatic',\n",
    "    'UAC' : 'aromatic',\n",
    "    'UGG' : 'aromatic'\n",
    "\n",
    "}\n",
    "\n",
    "print('CODE GENERATED BY:')\n",
    "\n",
    "print(\"     _    _ _                  __  __       _   _                   \")\n",
    "print(\"    / \\\\  | (_)_ __   __ _     |  \\\\/  | __ _| |_| |__   _____      __\")\n",
    "print(\"   / _ \\\\ | | | '_ \\\\ / _` |    | |\\\\/| |/ _` | __| '_ \\\\ / _ \\\\ \\\\ /\\\\ / /\")\n",
    "print(\"  / ___ \\\\| | | | | | (_| |    | |  | | (_| | |_| | | |  __/\\\\ V  V / \")\n",
    "print(\" /_/   \\\\_\\\\_|_|_| |_|\\\\__,_|    |_|  |_|\\\\__,_|\\\\__|_| |_|\\\\___| \\\\_/\\\\_/  \")\n",
    "print(\"                                                                    \")\n",
    "\n",
    "print()\n",
    "\n",
    "print(\"CODONS TO AMINO ACIDS: LOADING COMPLETE!\")\n",
    "print(\"PROPERTIES OF AMINO ACIDS: LOADING COMPLETE!\")\n",
    "print()\n",
    "print('  _________\\n /         \\\\\\n |  /\\\\ /\\\\  |\\n |    -    |\\n |  \\\\___/  |\\n \\\\_________/')\n",
    "\n",
    "print()\n",
    "\n",
    "DNA_length = int(input(\"PROVIDE DESIRED LENGTH OF DNA: \"))\n",
    "print()\n",
    "\n",
    "PAUSE = 0.15  # Change it to adjust the animation speed\n",
    "\n",
    "\n",
    "\n",
    "try:\n",
    "\n",
    "    # Generate random DNA strings\n",
    "    original_DNA = generate_random_DNA_string(DNA_length)\n",
    "    mutated_DNA = original_DNA[:]   #copies the original DNA so then later we can introduce mutation in a few base pairs\n",
    "\n",
    "    # Introduce mutations in a few base pairs\n",
    "    num_mutations = random.randint(1, 5)\n",
    "    for _ in range(num_mutations):\n",
    "        mutation_position = random.randint(0, len(mutated_DNA) - 1)\n",
    "        mutated_base = random.choice('ATCG'.replace(mutated_DNA[mutation_position], ''))\n",
    "        mutated_DNA = mutated_DNA[:mutation_position] + mutated_base + mutated_DNA[mutation_position + 1:]\n",
    "\n",
    "    print(\"ORIGINAL DNA STRING:\", original_DNA)\n",
    "    print()\n",
    "    print(\"MUTATED DNA STRING:\", mutated_DNA)\n",
    "    print()\n",
    "\n",
    "    compare = Compare_Codons(original_DNA, mutated_DNA)\n",
    "    RNA1, RNA2 = compare.RNA_strings()\n",
    "    compare.compare_RNA_strings(RNA1, RNA2)\n",
    "    compare.compare_codons(RNA1, RNA2)\n",
    "\n",
    "    # Find point mutations and visualize them\n",
    "    mutations = find_point_mutations(RNA1, RNA2)\n",
    "    visualize_mutations(original_DNA, mutated_DNA, mutations)\n",
    "except KeyboardInterrupt:\n",
    "    sys.exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42fcf934",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
